README.md
# exegesis-plugin-google-oauth2

[![Run Status](https://api.shippable.com/projects/5c32b19e302eb707003c01b6/badge?branch=master)]()
[![Coverage Badge](https://api.shippable.com/projects/5c32b19e302eb707003c01b6/coverageBadge?branch=master)]()
![](https://img.shields.io/github/issues/phil-mitchell/exegesis-plugin-google-oauth2.svg)
![](https://img.shields.io/github/license/phil-mitchell/exegesis-plugin-google-oauth2.svg)
![](https://img.shields.io/node/v/exegesis-plugin-google-oauth2.svg)
![](https://img.shields.io/npm/dependency-version/exegesis-plugin-google-oauth2/googleapis.svg)

## Description

Exegesis middleware to handle Google OAuth2 authentication

## Installation

```sh
npm install exegesis-plugin-google-oauth2
```

## Example

Add this to your Exegesis options:

```js
const exegesisGoogleOAuth2Plugin = require( 'exegesis-plugin-google-oauth2' );

options = {
    plugins: [
        exegesisGoogleOAuth2Plugin({
            // URL path to expose authentication endpoints (default /auth/google)
            path: '/auth/google',

            // Client ID and secret for Google API
            clientId: 'myGoogleClient',
            clientSecret: 'MyClientSecret',

            // Authorization parameters to be passed to Google API (defaults shown)
            authorization: {
                access_type: 'online',
                prompt: 'consent',
                scope: [ 'email', 'profile' ]
            },

            // Options to pass to people.get (see https://developers.google.com/people/api/rest/v1/people/get)
            people: {
                skip: false, // don't call the people API
                resourceName: 'people/me',
                personFields: 'names,nicknames,coverPhotos,emailAddresses'
            },

            // Callback function to handle successful authentication
            // parameters:
            // - context: exegesis context object
            // - tokens: tokens received from Google
            // - me: Google People user profile
            callback: async function( context, tokens, me ) {
                return{
                    id: me.id
                }
            }
        })
    ]
};
```

The environment variables `GOOGLE_CLIENT_ID` and `GOOGLE_CLIENT_SECRET` may also be used to provide the clientId and clientSecret.

If no callback function is provided, the response to successful authentication will be a JSON object:

```
{
    id: me.id,
    name: me.displayName,
    emails: me.emails,
    image: me.image
}
```


CHANGELOG.md
## [1.0.9](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/compare/v1.0.8...v1.0.9) (2021-05-09)


### Bug Fixes

* **deps:** Update dependencies ([589c776](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/commit/589c776f0b0e3c2c916ea74a147c179643b5156d))

## [1.0.8](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/compare/v1.0.7...v1.0.8) (2021-01-20)


### Bug Fixes

* **redirect:** parse the Forwarded header to determine the redirect url ([#16](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/issues/16)) ([ee3affc](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/commit/ee3affcc8b468cdbf9d9e61714fe8e2e5b7367cc))

## [1.0.7](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/compare/v1.0.6...v1.0.7) (2020-12-21)


### Bug Fixes

* **dependencies:** update dependencies to resolve vulnerabilities ([#15](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/issues/15)) ([26465c5](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/commit/26465c50ec0c4a716d6b62e76da813742ab0f076))

## [1.0.6](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/compare/v1.0.5...v1.0.6) (2020-02-02)


### Bug Fixes

* **me:** Fetch the photos field ([#8](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/issues/8)) ([c1527cc](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/commit/c1527cc8df1ee044fc3f16c442ee765dd183b774))

## [1.0.5](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/compare/v1.0.4...v1.0.5) (2020-02-01)


### Bug Fixes

* **deps:** update dependencies ([#7](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/issues/7)) ([ce2e293](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/commit/ce2e293b456eb618c1ff95926abe417d2bd88b9e))

## [1.0.4](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/compare/v1.0.3...v1.0.4) (2019-08-17)


### Bug Fixes

* **dependencies:** update dependencies ([593af2d](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/commit/593af2d))

## [1.0.3](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/compare/v1.0.2...v1.0.3) (2019-02-02)


### Bug Fixes

* **api:** Provide necessary parameters to the profile API ([8afa5b3](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/commit/8afa5b3))

## [1.0.2](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/compare/v1.0.1...v1.0.2) (2019-02-02)


### Bug Fixes

* **api:** Use the people API to get the user profile. ([586e21e](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/commit/586e21e))

## [1.0.1](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/compare/v1.0.0...v1.0.1) (2019-01-07)


### Bug Fixes

* **release:** fix title of README ([43c2ff5](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/commit/43c2ff5))

# 1.0.0 (2019-01-07)


### Bug Fixes

* **release:** initial release ([e6d362e](https://github.com/phil-mitchell/exegesis-plugin-google-oauth2/commit/e6d362e))


test/app.js
'use strict';

const express = require( 'express' );
const exegesisExpress = require( 'exegesis-express' );
const http = require( 'http' );
const path = require( 'path' );
const fs = require( 'fs-extra' );

const require_helper = require( './require_helper' );

var google = require( 'googleapis' ).google;
var GoogleOAuth2 = google.auth.OAuth2;
google.auth.OAuth2 = class MockOAuth2 extends GoogleOAuth2 {
    constructor() {
        super( ...arguments );
        this.getToken = async function( code ) {
            if( code === 'BadCode' ) {
                throw new Error( 'Bad code' );
            }
            return{
                tokens: {
                    access_token: code,
                    scope: 'https://www.googleapis.com/auth/userinfo.email',
                    token_type: 'Bearer',
                    id_token: 'idToken1',
                    expiry_date: new Date().getTime() + 50
                }
            };
        };
        this.setCredentials = function( tokens ) {
            this.tokens = tokens;
        };
        this.requestAsync = async function( options ) {
            if( this.tokens.access_token === 'BadToken' ) {
                throw new Error( 'Bad token' );
            }
            var data = {
                id: 'testid1',
                name: 'Test User',
                emails: [ 'test@google.com' ],
                image: 'http://localhost/image.png'
            };
            if( options.params.personFields === 'names' ) {
                delete data.emails;
                delete data.image;
            }
            return{
                status: this.tokens.access_token === 'FailedRequest' ? 400 : 200,
                data: data
            };
        };
    }
};

async function authCallback1( context, tokens, me ) {
    if( tokens.access_token === 'authcode1' ) {
        return{
            token: me.id + ':' + tokens.access_token
        };
    }

    throw context.makeError( 403, 'Not authorized for user ' + me.id );
}

async function createServer() {
    const app = express();

    const options = {
        controllers: {},
        allowMissingControllers: true,
        plugins: [
            require_helper( 'index.js' )({
                clientId: 'asdf',
                clientSecret: 'fdsa'
            }),
            require_helper( 'index.js' )({
                clientId: 'asdf',
                clientSecret: 'fdsa',
                path: '/myauthtest',
                callback: authCallback1
            }),
            require_helper( 'index.js' )({
                clientId: 'asdf',
                clientSecret: 'fdsa',
                path: '/skippeople',
                people: { skip: true }
            }),
            require_helper( 'index.js' )({
                clientId: 'asdf',
                clientSecret: 'fdsa',
                path: '/personfields',
                people: { personFields: 'names' }
            })
        ]
    };

    const exegesisMiddleware = await exegesisExpress.middleware(
        path.resolve( __dirname, './petstore.yaml' ),
        options
    );

    app.use( exegesisMiddleware );
    // Return a 404
    app.use( ( req, res ) => {
        res.status( 404 ).json({ message: `Not found` });
    });

    // Handle any unexpected errors
    app.use( ( err, req, res, next ) => {
        res.status( 500 ).json({ message: `Internal error: ${err.message}` });
        next();
    });

    const server = http.createServer( app );

    return server;
}

if( process.env.REPORT_DIR_FOR_CODE_COVERAGE ) {
    const dumpCoverage = () => {
        console.warn( 'Outputting code coverage information to ' + process.env.REPORT_DIR_FOR_CODE_COVERAGE );
        fs.ensureDirSync( process.env.REPORT_DIR_FOR_CODE_COVERAGE );
        fs.writeFileSync(
            process.env.REPORT_DIR_FOR_CODE_COVERAGE + '/app.json',
            JSON.stringify( global['__coverage__'] ), 'utf8'
        );
    };
    process.on( 'exit', dumpCoverage );
    process.on( 'SIGINT', process.exit );
    process.on( 'SIGTERM', process.exit );
}

createServer()
.then( server => {
    server.listen( 3000 );
    console.log( 'Listening on port 3000' );
    console.log( 'Try visiting http://localhost:3000/api-docs' );
})
.catch( err => {
    console.error( err.stack );
    process.exit( 1 );
});


test/unit/index.spec.js
'use strict';

var expect = require( 'chai' ).expect;
var require_helper = require( '../require_helper' );
var index = require_helper( 'index' );

describe( 'openapi version check', function() {
    before( function() {
        this.plugin = index({});
    });
    it( 'throws an error if there is no openapi', function() {
        expect( this.plugin.makeExegesisPlugin.bind(
            this.plugin, {
                apiDoc: {
                }
            }) ).to.throw( "OpenAPI definition is missing 'openapi' field" );
    });
    it( 'throws an error if openapi is too low', function() {
        expect( this.plugin.makeExegesisPlugin.bind(
            this.plugin, {
                apiDoc: {
                    openapi: '2.0.0'
                }
            }) ).to.throw( 'OpenAPI version 2.0.0 not supported' );
    });
    it( 'throws an error if openapi is too high', function() {
        expect( this.plugin.makeExegesisPlugin.bind(
            this.plugin, {
                apiDoc: {
                    openapi: '4.0.0'
                }
            }) ).to.throw( 'OpenAPI version 4.0.0 not supported' );
    });
});

describe( 'Google client id + secret from environment', function() {
    before( function() {
        delete process.env.GOOGLE_CLIENT_ID;
        delete process.env.GOOGLE_CLIENT_SECRET;
        this.plugin = index({});
    });
    after( function() {
        delete process.env.GOOGLE_CLIENT_ID;
        delete process.env.GOOGLE_CLIENT_SECRET;
    });
    it( 'throws an error if no variables are set', function() {
        delete process.env.GOOGLE_CLIENT_ID;
        delete process.env.GOOGLE_CLIENT_SECRET;
        expect( this.plugin.makeExegesisPlugin.bind(
            this.plugin, {
                apiDoc: {
                    openapi: '3.0.0'
                }
            }) ).to.throw( 'options requires clientId and clientSecret for Google API access' );
    });
    it( 'throws an error if only GOOGLE_CLIENT_ID is set', function() {
        process.env.GOOGLE_CLIENT_ID='test1';
        delete process.env.GOOGLE_CLIENT_SECRET;
        expect( this.plugin.makeExegesisPlugin.bind(
            this.plugin, {
                apiDoc: {
                    openapi: '3.0.0'
                }
            }) ).to.throw( 'options requires clientId and clientSecret for Google API access' );
    });
    it( 'throws an error if only GOOGLE_CLIENT_SECRET is set', function() {
        delete process.env.GOOGLE_CLIENT_ID;
        process.env.GOOGLE_CLIENT_SECRET='secret1';
        expect( this.plugin.makeExegesisPlugin.bind(
            this.plugin, {
                apiDoc: {
                    openapi: '3.0.0'
                }
            }) ).to.throw( 'options requires clientId and clientSecret for Google API access' );
    });
    it( 'succeeds if both are set', function() {
        process.env.GOOGLE_CLIENT_ID='test1';
        process.env.GOOGLE_CLIENT_SECRET='secret1';
        let instance = this.plugin.makeExegesisPlugin({
            apiDoc: {
                openapi: '3.0.0'
            }
        });
        expect( instance ).to.be.ok;
        expect( instance._options.clientId ).to.equal( 'test1' );
        expect( instance._options.clientSecret ).to.equal( 'secret1' );
    });
});

describe( 'Google client id + secret from options', function() {
    before( function() {
        delete process.env.GOOGLE_CLIENT_ID;
        delete process.env.GOOGLE_CLIENT_SECRET;
        this.plugin = index({
            clientId: 'test2',
            clientSecret: 'secret2'
        });
    });
    after( function() {
        delete process.env.GOOGLE_CLIENT_ID;
        delete process.env.GOOGLE_CLIENT_SECRET;
    });
    it( 'succeeds without environment variables', function() {
        delete process.env.GOOGLE_CLIENT_ID;
        delete process.env.GOOGLE_CLIENT_SECRET;
        let instance = this.plugin.makeExegesisPlugin({
            apiDoc: {
                openapi: '3.0.0'
            }
        });
        expect( instance ).to.be.ok;
        expect( instance._options.clientId ).to.equal( 'test2' );
        expect( instance._options.clientSecret ).to.equal( 'secret2' );
    });
    it( 'does not override with environment variables', function() {
        process.env.GOOGLE_CLIENT_ID='test1';
        process.env.GOOGLE_CLIENT_SECRET='secret1';
        let instance = this.plugin.makeExegesisPlugin({
            apiDoc: {
                openapi: '3.0.0'
            }
        });
        expect( instance ).to.be.ok;
        expect( instance._options.clientId ).to.equal( 'test2' );
        expect( instance._options.clientSecret ).to.equal( 'secret2' );
    });
});

describe( 'No path in options', function() {
    before( function() {
        delete process.env.GOOGLE_CLIENT_ID;
        delete process.env.GOOGLE_CLIENT_SECRET;
        this.plugin = index({
            clientId: 'test2',
            clientSecret: 'secret2'
        });
        this.apiDoc = {
            openapi: '3.0.0'
        };
        this.controllers = {};
        this.instance = this.plugin.makeExegesisPlugin({ apiDoc: this.apiDoc });
        this.instance.preCompile({ apiDoc: this.apiDoc, options: {
            controllers: this.controllers
        } });
    });
    it( 'uses /auth/google as default path', function() {
        expect( this.apiDoc.paths ).to.be.ok;
        expect( this.apiDoc.paths['/auth/google'] ).to.be.ok;
        expect( this.apiDoc.paths['/auth/google'].get ).to.be.ok;
    });
    it( 'creates controller function', function() {
        expect( this.controllers[this.apiDoc.paths['/auth/google'].get['x-exegesis-controller']] ).to.be.ok;
        expect( this.controllers[this.apiDoc.paths['/auth/google'].get['x-exegesis-controller']][
        this.apiDoc.paths['/auth/google'].get.operationId] ).to.be.a( 'function' );
    });
    it( 'sends a redirect to the correct path', async function() {
        var context = {
            req: {
                url: '/auth/google',
                protocol: 'proto',
                get: function( field ) {
                    if( field === 'host' ) {
                        return( 'myhost1' );
                    }
                }
            },
            res: {
                setStatus: function( status ) {
                    this.status = status;
                    return this;
                },
                set: function( header, value ) {
                    if( header === 'location' ) {
                        this.location = value;
                    }
                }
            },
            params: {
            }
        };
        await this.controllers[this.apiDoc.paths['/auth/google'].get['x-exegesis-controller']][
        this.apiDoc.paths['/auth/google'].get.operationId]( context );
        expect( context.res.status ).to.equal( 303 );
        expect( context.res.location ).to.contain(
            'redirect_uri=' + encodeURIComponent( 'proto://myhost1/auth/google' ) );
    });
});

describe( 'custom path in options', function() {
    before( function() {
        delete process.env.GOOGLE_CLIENT_ID;
        delete process.env.GOOGLE_CLIENT_SECRET;
        this.plugin = index({
            clientId: 'test2',
            clientSecret: 'secret2',
            path: '/my/path/to/auth'
        });
        this.apiDoc = {
            openapi: '3.0.0'
        };
        this.controllers = {};
        this.instance = this.plugin.makeExegesisPlugin({ apiDoc: this.apiDoc });
        this.instance.preCompile({ apiDoc: this.apiDoc, options: {
            controllers: this.controllers
        } });
    });
    it( 'uses specified path', function() {
        expect( this.apiDoc.paths ).to.be.ok;
        expect( this.apiDoc.paths['/my/path/to/auth'] ).to.be.ok;
        expect( this.apiDoc.paths['/my/path/to/auth'].get ).to.be.ok;
    });
    it( 'creates controller function', function() {
        expect( this.controllers[this.apiDoc.paths['/my/path/to/auth'].get['x-exegesis-controller']] )
        .to.be.ok;
        expect( this.controllers[this.apiDoc.paths['/my/path/to/auth'].get['x-exegesis-controller']][
        this.apiDoc.paths['/my/path/to/auth'].get.operationId] ).to.be.a( 'function' );
    });
});


test/integration/options.spec.js
'use strict';

var superagent = require( 'superagent' );
var expect = require( 'chai' ).expect;

describe( 'Default options', function() {
    it( 'redirects google oauth', function( done ) {
        superagent
        .get( 'http://localhost:3000/auth/google' )
        .redirects( 0 )
        .ok( res => res.status === 303 )
        .end( function( e, res ) {
            expect( e ).to.be.null;
            expect( res.headers.location ).to.contain( 'https://accounts.google.com/o/oauth2/v2/auth?' );
            expect( res.headers.location ).to.contain(
                'redirect_uri=' + encodeURIComponent( 'http://localhost:3000/auth/google' ) );
            done();
        });
    });

    it( 'handles the code from google', function( done ) {
        superagent
        .get( 'http://localhost:3000/auth/google' )
        .query({ code: 'authcode1' })
        .ok( res => res.status === 200 )
        .end( function( e, res ) {
            expect( e ).to.be.null;
            expect( res.body.id ).to.equal( 'testid1' );
            expect( res.body.emails ).to.contain( 'test@google.com' );
            expect( res.body.image ).to.equal( 'http://localhost/image.png' );
            done();
        });
    });

    it( 'handles bad code from Google', function( done ) {
        superagent
        .get( 'http://localhost:3000/auth/google' )
        .query({ code: 'BadCode' })
        .ok( res => res.status === 500 )
        .end( function( e, res ) {
            expect( e ).to.be.null;
            expect( res.body.message ).to.equal( 'Bad code' );
            done();
        });
    });

    it( 'handles bad token from Google', function( done ) {
        superagent
        .get( 'http://localhost:3000/auth/google' )
        .query({ code: 'BadToken' })
        .ok( res => res.status === 500 )
        .end( function( e, res ) {
            expect( e ).to.be.null;
            expect( res.body.message ).to.equal( 'Bad token' );
            done();
        });
    });

    it( 'handles failed request to Google', function( done ) {
        superagent
        .get( 'http://localhost:3000/auth/google' )
        .query({ code: 'FailedRequest' })
        .ok( res => res.status === 500 )
        .end( function( e, res ) {
            expect( e ).to.be.null;
            expect( res.body.message ).to.equal( 'Failed to get user information from Google' );
            done();
        });
    });
});

describe( 'Custom path and callback', function() {
    it( 'handles the code from google', function( done ) {
        superagent
        .get( 'http://localhost:3000/myauthtest' )
        .query({ code: 'authcode1' })
        .ok( res => res.status === 200 )
        .end( function( e, res ) {
            expect( e ).to.be.null;
            expect( res.body.token ).to.equal( 'testid1:authcode1' );
            done();
        });
    });

    it( 'handles bad code from Google', function( done ) {
        superagent
        .get( 'http://localhost:3000/myauthtest' )
        .query({ code: 'authcode2' })
        .ok( res => res.status === 403 )
        .end( function( e, res ) {
            expect( e ).to.be.null;
            expect( res.body.message ).to.contain( 'Not authorized for user testid1' );
            done();
        });
    });
});

describe( 'Skip people API', function() {
    it( 'handles the code from google', function( done ) {
        superagent
        .get( 'http://localhost:3000/skippeople' )
        .query({ code: 'authcode1' })
        .ok( res => res.status === 200 )
        .end( function( e, res ) {
            expect( e ).to.be.null;
            expect( res.body.access_token ).to.eql( 'authcode1' );
            done();
        });
    });
});

describe( 'Override personFields', function() {
    it( 'handles the code from google', function( done ) {
        superagent
        .get( 'http://localhost:3000/personfields' )
        .query({ code: 'authcode1' })
        .ok( res => res.status === 200 )
        .end( function( e, res ) {
            expect( e ).to.be.null;
            expect( res.body.id ).to.equal( 'testid1' );
            expect( res.body.emails ).to.not.exist;
            expect( res.body.image ).to.not.exist;
            done();
        });
    });
});


gruntfile.js
'use strict';

const path = require( 'path' );

//const is_ci = process.env.CI === 'true';
const is_shippable = process.env.SHIPPABLE === 'true';

var coverage_dir = is_shippable ? './shippable/codecoverage' : './coverage';
var results_dir = is_shippable ? './shippable/testresults' : '.';

module.exports = function( grunt ) {
    grunt.initConfig({
        clean: [ 'coverage', 'shippable', 'xunit.xml' ],
        env: {
            coverage: {
                APP_DIR_FOR_CODE_COVERAGE: path.resolve( __dirname, 'coverage/instrument' ),
                REPORT_DIR_FOR_CODE_COVERAGE: path.resolve( __dirname, 'coverage/reports' )
            },
            unit: {
                XUNIT_FILE: results_dir + '/unittests.xml'
            },
            integration: {
                XUNIT_FILE: results_dir + '/integrationtests.xml'
            }
        },
        mkdir: {
            report: {
                options: {
                    create: [ path.dirname( process.env.XUNIT_FILE || './xunit.xml' ) ]
                }
            }
        },
        express: {
            test: {
                options: {
                    script: './test/app.js'
                }
            }
        },
        instrument: {
            files: 'src/**/*.js',
            options: {
                lazy: false,
                basePath: './coverage/instrument/'
            }
        },
        mochaTest: {
            options: {
                reporter: 'spec-xunit-file'
            },
            unit: {
                src: [ 'test/unit/**/*.spec.js' ]
            },
            integration: {
                src: [ 'test/integration/**/*.spec.js' ]
            }
        },
        storeCoverage: {
            options: {
                dir: './coverage/reports',
                'include-all-sources': true
            }
        },
        makeReport: {
            src: './coverage/reports/**/*.json',
            options: {
                type: [ 'cobertura', 'html', 'json', 'lcovonly' ],
                dir: coverage_dir,
                print: 'detail'
            }
        }
    });
    grunt.loadNpmTasks( 'grunt-express-server' );
    grunt.loadNpmTasks( 'grunt-mocha-test' );
    grunt.loadNpmTasks( 'grunt-istanbul' );
    grunt.loadNpmTasks( 'grunt-env' );
    grunt.loadNpmTasks( 'grunt-contrib-clean' );
    grunt.loadNpmTasks( 'grunt-mkdir' );

    grunt.registerTask( 'dummyCoverage', () => {
        global['__coverage__'] = global['__coverage__'] || {};
    });

    grunt.registerTask( 'unit', [ 'env:unit', 'mkdir:report', 'mochaTest:unit' ] );
    grunt.registerTask( 'integration', [ 'env:integration', 'express:test', 'mkdir:report', 'mochaTest:integration', 'express:test:stop' ] );
    grunt.registerTask( 'default', [ 'clean', 'unit', 'integration' ] );
    grunt.registerTask( 'coverage', [ 'clean', /*'env:coverage', 'instrument',*/
                                      'unit', 'integration', /*'dummyCoverage',
                                      'storeCoverage', 'makeReport'*/ ] );
};


src/index.js
'use strict';

const semver = require( 'semver' );
const google = require( 'googleapis' ).google;
const url = require( 'url' );

const controllerName = 'exegesis-plugin-google-oauth2';

class GoogleOAuth2Plugin {
    constructor( apiDoc, options ) {
        // Verify the apiDoc is an OpenAPI 3.x.x document, because this plugin
        // doesn't know how to handle anything else.
        if( !apiDoc.openapi ) {
            throw new Error( "OpenAPI definition is missing 'openapi' field" );
        }

        if( !semver.satisfies( apiDoc.openapi, '>=3.0.0 <4.0.0' ) ) {
            throw new Error( `OpenAPI version ${apiDoc.openapi} not supported` );
        }

        options = Object.assign({
            clientId: process.env.GOOGLE_CLIENT_ID,
            clientSecret: process.env.GOOGLE_CLIENT_SECRET
        }, options );

        if( !options.clientId || !options.clientSecret ) {
            throw new Error( 'options requires clientId and clientSecret for Google API access' );
        }

        options.authorization = Object.assign({
            access_type: 'online',
            prompt: 'consent',
            scope: [ 'email', 'profile' ]
        }, options.authorization );
        options.path = options.path || '/auth/google';

        this._options = options;

        apiDoc.paths = apiDoc.paths || {};
        apiDoc.paths[options.path] = {
            'get': {
                summary: 'Authenticate with Google OAuth2',
                'x-exegesis-controller': controllerName,
                operationId: 'authenticate ' + options.path,
                security: [],
                parameters: [ {
                    'in': 'query',
                    name: 'code',
                    description: 'The authentication code provided by Google',
                    schema: {
                        type: 'string'
                    }
                } ],
                responses: {
                    '200': {
                        description: 'Authentication was successful'
                    },
                    '303': {
                        description: 'Redirect to start OAuth handshake'
                    }
                }
            }
        };
    }

    preCompile({ options }) {
        var authOptions = this._options;

        options.controllers[controllerName] = options.controllers[controllerName] || {};
        options.controllers[controllerName]['authenticate ' + authOptions.path ] = async function( context ) {
            var parsedUrl = url.parse( context.req.url );
            var protocol = context.req.protocol;
            var host = context.req.get( 'host' );
            var forwarded = context.req.get( 'forwarded' );
            if( forwarded ) {
                for( let field of forwarded.split( ';' ) ) {
                    let parts = field.split( '=' );
                    if( parts[0] === 'proto' ) {
                        protocol = parts.slice( 1 ).join( '=' );
                    } else if( parts[0] === 'host' ) {
                        host = parts.slice( 1 ).join( '=' );
                    }
                }
            }
            var auth = new google.auth.OAuth2(
                authOptions.clientId,
                authOptions.clientSecret,
                protocol + '://' + host + parsedUrl.pathname
            );

            var query = context.params.query;
            if( !query || !query.code ) {
                context.res.setStatus( 303 ).set( 'location', auth.generateAuthUrl( JSON.parse( JSON.stringify( authOptions.authorization ) ) ) );
            } else {
                let me = null;
                let tokens = null;
                try {
                    let data = await auth.getToken( query.code );
                    tokens = data.tokens;
                    if( !authOptions.people || !authOptions.people.skip ) {
                        auth.setCredentials( tokens );
                        let people = google.people({ version: 'v1', auth });
                        me = await people.people.get( Object.assign({
                            resourceName: 'people/me',
                            personFields: 'names,nicknames,photos,coverPhotos,emailAddresses'
                        }, authOptions.people || {}) );
                        if( me.status !== 200 || !me.data ) {
                            throw new Error( 'Failed to get user information from Google' );
                        }
                        me = me.data;
                    }
                } catch( e ) {
                    return context.res.setStatus( 500 ).json({
                        message: e.message
                    });
                }
                if( typeof( authOptions.callback ) === 'function' ) {
                    return authOptions.callback( context, tokens, me );
                }
                if( me ) {
                    return context.res.json({
                        id: me.id,
                        name: me.displayName,
                        emails: me.emails,
                        image: me.image
                    });
                }
                return context.res.json( tokens );
            }
        };
    }
}

module.exports = function plugin( options ) {
    return{
        info: {
            name: 'exegesis-plugin-google-oauth2'
        },
        options: options,
        makeExegesisPlugin: function makeExegesisPlugin({ apiDoc }) {
            return new GoogleOAuth2Plugin( apiDoc, options );
        }
    };
};
